mirgecom/euler.py

```diff
@@ -20,20 +22,335 @@
+    """
+    mass: np.ndarray
+    energy: np.ndarray
+    momentum: np.ndarray
+
+
+@dataclass
+class MassFractions:
+    r"""
+    Class to pick off the species mass fractions
+    (mass fractions) per unit volume =
+    :math:`(\rhoY_{\alpha}) | 1 \le \alpha \le N_{species}`,
+    from an agglomerated object array. :math:`N_{species}` is
+    the number of mixture species.
+
+    .. attribute:: mass
+
+        Mass fraction per unit volume for each mixture species
+    """
+    massfractions: np.ndarray
+
+
+def split_fields(ndim, q):
+    """
+    Method to spit out a list of named flow variables in
+    an agglomerated flow solution. Useful for specifying
+    named data arrays to helper functions (e.g. I/O).
+    """
+    qs = split_conserved(ndim, q)
+    mass = qs.mass
+    energy = qs.energy
+    mom = qs.momentum
+
+    retlist = [
+        ("mass", mass),
+        ("energy", energy),
+        ("momentum", mom),
+    ]
+    nscalar = number_of_scalars(ndim, q)
+    if nscalar > 0:
+        massfrac = split_species(ndim, q).massfraction
+        retlist.append(("massfraction", massfrac))
+
+    return retlist
+
+
+def number_of_scalars(ndim, q):
+    """
+    Return the number of scalars or mixture species in a flow solution.
+    """
+    return len(q) - (ndim + 2)
+
+
+def number_of_equations(ndim, q):
+    """
+    Return the number of equations (i.e. number of dofs) in the soln
+    """
+    return len(q) + number_of_scalars(ndim, q)
+
+
+def split_conserved(dim, q):
+    """
+    Return a 'ConservedVars' object that splits out conserved quantities
+    by name. Useful for expressive coding.
+    """
+    return ConservedVars(mass=q[0], energy=q[1], momentum=q[2:2+dim])
+
+
+def split_species(dim, q):
+    """
+    Return a 'MassFractions' object that splits out mixture species
+    conserved quantities by name. Useful for expressive coding.
+    """
+    numscalar = number_of_scalars(dim, q)
+    sindex = dim + 2
+    return MassFractions(massfractions=q[sindex:sindex+numscalar])
+
+
+def _inviscid_flux(discr, q, eos=IdealSingleGas()):
+    r"""Computes the inviscid flux vectors from flow solution *q*
+
+    The inviscid fluxes are
+    :math:`(\rho\vec{V},(\rhoE+p)\vec{V},\rho(\vec{V}\otimes\vec{V})+p\mathbf{I})
+    """
+    ndim = discr.dim
+
+    # q = [ rho rhoE rhoV ]
+    qs = split_conserved(ndim, q)
+    mass = qs.mass
+    energy = qs.energy
+    mom = qs.momentum
+
+    p = eos.pressure(q)
+
+    # Fluxes:
+    # [ rhoV (rhoE + p)V (rhoV.x.V + p*I) ]
+    momflux = make_obj_array(
+        [
+            (mom[i] * mom[j] / mass + (p if i == j else 0))
+            for i in range(ndim)
+            for j in range(ndim)
+        ]
+    )
+    massflux = mom * make_obj_array([1.0])
+    energyflux = mom * make_obj_array([(energy + p) / mass])
+    # scalarflux = mom * massfractions / mass
+
+    return flat_obj_array(massflux, energyflux, momflux,)
+
+
+def _get_wavespeed(dim, q, eos=IdealSingleGas()):
```

Comment:
I'm not sure I like that `eos` has a default. I don't think it's ever used without one. In addition, I'd like the argument order be such that the state vector is last.